src/apps/btc/btc_sign_msg.c

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "btc_sign_msg.h"
+#include "btc_common.h"
+
+#include <hardfault.h>
+#include <hww.pb.h>
+#include <keystore.h>
+#include <rust/rust.h>
+#include <workflow/confirm.h>
+
+#include <wally_script.h>
+
+#include <stdio.h>
+
+#define MAX_MESSAGE_SIZE 1024
+
+USE_RESULT app_btc_result_t app_btc_sign_msg(
+    BTCCoin coin,
+    const BTCScriptConfig* script_config,
+    const uint32_t* keypath,
+    size_t keypath_len,
+    const uint8_t* msg,
+    size_t msg_size,
+    uint8_t* signature_out)
+{
+    if (coin != BTCCoin_BTC) {
+        return APP_BTC_ERR_INVALID_INPUT;
+    }
+    if (script_config->which_config != BTCScriptConfig_simple_type_tag) {
+        return APP_BTC_ERR_INVALID_INPUT;
+    }
+    if (msg_size > MAX_MESSAGE_SIZE) {
+        return APP_BTC_ERR_INVALID_INPUT;
+    }
+    // Keypath and script_config are validated in app_btc_address_simple() below.
+
+    char basic_info[100] = {0};
+    int snprintf_result =
+        snprintf(basic_info, sizeof(basic_info), "Coin: %s", btc_common_coin_name(coin));
+    if (snprintf_result < 0 || snprintf_result >= (int)sizeof(basic_info)) {
+        Abort("app_btc_sign_message/0");
+    }
+    const confirm_params_t params_basic = {
+        .title = "Sign message",
+        .body = basic_info,
+        .accept_is_nextarrow = true,
+    };
+    if (!workflow_confirm_blocking(&params_basic)) {
+        return APP_BTC_ERR_USER_ABORT;
+    }
+    char address[500] = {0};
+    if (!app_btc_address_simple(
+            coin,
+            script_config->config.simple_type,
+            keypath,
+            keypath_len,
+            address,
+            sizeof(address))) {
+        return APP_BTC_ERR_INVALID_INPUT;
+    }
+    const confirm_params_t params_address = {
+        .title = "Address",
+        .body = address,
+        .scrollable = true,
+        .accept_is_nextarrow = true,
+    };
+    if (!workflow_confirm_blocking(&params_address)) {
+        return APP_BTC_ERR_USER_ABORT;
+    }
+
+    switch (rust_workflow_verify_message(rust_util_bytes(msg, msg_size))) {
+    case VerifyMessageResultInvalidInput:
+        return APP_BTC_ERR_INVALID_INPUT;
+    case VerifyMessageResultUserAbort:
+        return APP_BTC_ERR_USER_ABORT;
+    default:
+        break;
+    }
+
+    const char msg_header[] =
+        "\x18"
```

Comment:
A comment with a link where to look for Electrum format would've been very nice.

Reply:
added

Reply:
very nice. thanks! I'm missing one part though. in electrum, it creates this:

```python
return b"\x18Bitcoin Signed Message:\n" + length + message
```

the magic header + length of the message + the message itself, but I can't see where do we encode the "length of the message" part in our code here?

Reply:
ah never mind, it's in the `varint_to_bytes` in libwally-core.